File: Number of Unique Paths - GFG/number-of-unique-paths.java
//{ Driver Code Starts
//Initial Template for Java

/*package whatever //do not write package name here */

import java.io.*;
import java.util.*;

class GFG {
	public static void main (String[] args) {
	    
	    //taking input using Scanner class
		Scanner sc=new Scanner(System.in);
		
		//taking total testcases
		int t=sc.nextInt();
		while(t-->0)
		{
		    //taking dimensions of the matrix
		    int a=sc.nextInt();
		    int b=sc.nextInt();
		    Solution ob = new Solution();
		    //calling method NumberOfPath()
		    System.out.println(ob.NumberOfPath(a,b));
		}
	}
}
// } Driver Code Ends


//User function Template for Java

class Solution
{
    //Function to find total number of unique paths.
    public static int NumberOfPath(int n, int m) 
    {
        //Your code here
        int dp[][] = new int[n+1][m+1];
        dp[0][1] =1;
        for(int i=1; i<n+1; i++){
            for(int j=1; j<m+1; j++){
                dp[i][j] = dp[i-1][j]+dp[i][j-1];
            }
        }
        return dp[n][m];
        
    }
}

File: Number of Unique Paths - GFG/number-of-unique-paths.cpp
//{ Driver Code Starts
//Initial template for C++
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function template in C++

class Solution
{
    public:
     int dfs(int i,int j,int m,int n,vector<vector<int>>&dp){
        if(i>=m or j>=n ){
            return 0;
        }
        if(dp[i][j]!=-1){
            return dp[i][j];
        }
        if(i==m-1 and j==n-1){
            return 1;
        }
        return dp[i][j]=dfs(i+1,j,m,n,dp)+dfs(i,j+1,m,n,dp);
    }
    //Function to find total number of unique paths.
    int NumberOfPath(int m, int n)
    {
        vector<vector<int>>dp(m,vector<int>(n,-1));
        return dfs(0,0,m,n,dp);

    }
};


//{ Driver Code Starts.
int main()
{
    //taking total testcases
    int t;
    cin>>t;
    while(t--)
    {
        //taking dimensions of the matrix
        int a,b;
        cin>>a>>b;
        Solution ob;
        //calling NumberOfPath() function
        cout << ob.NumberOfPath(a,b) << endl;
    }
}


// } Driver Code Ends